### PR TITLE
feat: memory storage

### PR DESCRIPTION
**Pull request by mmkal**
BREAKING CHANGE: storage instances now must be passed into the Umzug constructor. strings (`'json'`, `'sequelize'`, `'mongodb'` etc.), as well as `storageOptions`, are no longer supported.

This also changes folder structure slightly:

before:
```
src/storages/JSONStorage.ts
src/storages/MongoDBStorage.ts
...
```

after:
```
src/storage/json.ts
src/storage/mongodb.ts
...
```

And all built-in storages are exported from the package's `main` module, so they can be imported with (e.g.) `import { JSONStorage } from 'umzug'`

### CODE REVIEW

File: legacy-tests/fixtures/index.js [hunk 1]
```diff
@@ -32,11 +32,10 @@ describe('custom resolver', () => {
...
           nameFormatter: (path) => parse(path).name,
         },
-        storage: 'sequelize',
-        storageOptions: {
+        storage: new SequelizeStorage({
           path: this.storagePath,
-          sequelize: this.sequelize,
-        },
+            sequelize: this.sequelize,
+          }),
```
[the remark this answers is not in the document]
Comment (mmkal) [reply legacy-tests/fixtures/index.js:38]: The formatting lint rule (prettier) is turned off for `legacy-tests`, to avoid noise (there were a _lot_ of violations). I have a branch where I move most of these legacy tests to jest, which formats them as a side-effect. Will fix this one manually though.

File: src/storage/memory.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { UmzugStorage } from './contract';
+
+export const memoryStorage = (): UmzugStorage => {
+	let executed = [] as string[];
```
[the remark this answers is not in the document]
Comment (mmkal) [reply src/storage/memory.ts:4]: I agree, that would make sense as a rule. Have changed manually for now.

File: src/storage/memory.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+	return {
+		logMigration: async name => {
+			executed.push(name);
+		},
+		unlogMigration: async name => {
+			executed = executed.filter(n => n !== name);
+		},
```
[the remark this answers is not in the document]
Comment (mmkal) [reply src/storage/memory.ts:11]: I agree, but this is mirroring the implementation of JSONStorage. Let's make both throw before releasing v3. I think it should go in as a standalone PR though so it can be marked as a separate breaking change.

File: test/storage/memory.test.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { expectTypeOf } from 'expect-type';
+import { UmzugStorage, memoryStorage } from '../../src';
+
+describe('memoryStorage', () => {
+	test('type', () => {
+		expectTypeOf(memoryStorage).returns.toMatchTypeOf<UmzugStorage>();
+		// no additional properties:
+		expectTypeOf(memoryStorage).returns.toEqualTypeOf<UmzugStorage>();
```
[the remark this answers is not in the document]
Comment (mmkal) [reply test/storage/memory.test.ts:8]: Good call out - have just pushed [a release to make this clearer](https://github.com/mmkal/ts/releases/tag/expect-type%400.7.1) - [diff](https://github.com/mmkal/ts/compare/expect-type@0.7.0...expect-type@0.7.1). Taking from docs there:


`.toEqualTypeOf` succeeds for objects with different values, but the same type:

```typescript
expectTypeOf({a: 1}).toEqualTypeOf({a: 2})
```

`.toMatchTypeOf` checks that an object "matches" a type - that is, it has all the expected properties with correct types. This is similar to jest's `.toMatchObject`:

```typescript
expectTypeOf({a: 1, b: 1}).toMatchTypeOf({a: 1})
```

So basically this `.toEqualTypeOf` assertion says that `memoryStorage` should return `UmzugStorage`, with no extra properties.

File: src/storage/index.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+// codegen:start {preset: barrel}
+export * from './contract';
+export * from './json';
+export * from './memory';
+export * from './mongodb';
+export * from './sequelize';
+// codegen:end
```
[the remark this answers is not in the document]
Comment (mmkal) [reply src/storage/index.ts:7]: Not all that much! It scans the directory for other files and generates the `export` statements. If we rename/move/delete/add files it'll be a lint error, and it'll auto-fix itself to make sure this "barrel" stays in sync.

File: src/storages/type-helpers/sequelize-type-helpers.ts [hunk 6]
```diff
@@ -1,12 +0,0 @@
-import { Sequelize as SequelizeType, Model as ModelClass } from 'sequelize';
-
-interface ModelTempInterface extends ModelClass {
-	[key: string]: any;
-}
-
-type ModelClassType = typeof ModelClass & {
-	// eslint-disable-next-line @typescript-eslint/prefer-function-type
-	new (values?: object, options?: any): ModelTempInterface;
-};
```
[the remark this answers is not in the document]
Comment (mmkal) [reply src/storages/type-helpers/sequelize-type-helpers.ts:10]: I didn't! I just removed the `eslint-disable` and it auto fixed itself!

File: tsconfig.json [hunk 7]
```diff
@@ -15,8 +15,8 @@
 		"newLine": "lf",
 		"stripInternal": true,
 		"noImplicitReturns": true,
-		"noUnusedLocals": true,
-		"noUnusedParameters": true,
+		"noUnusedLocals": false,
+		"noUnusedParameters": false,
```
[the remark this answers is not in the document]
Comment (mmkal) [reply tsconfig.json:19]: IMO this is a linter concern, not a compiler one. It's duplicated my the `no-unused-variables` rule, and better via eslint because:

- if we do need to suppress it in one-off cases, it's safer to disable the specific rule with eslint rather than `// @ts-ignore`
- the eslint rule is more configurable. We can allow certain regexes to establish conventions
- ts-jest will refuse to run if there are compiler errors. This can be turned off but it's a useful feature. But when developing it's or pretty common to run tests with tempors unused variables. They won't get into master anyway because of the lint rule.

File: legacy-tests/fixtures/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (papb) [reply legacy-tests/fixtures/index.js]: Ah, indeed, I believe I actually suggested turning it off for legacy-tests myself :sweat_smile: ok!!

File: tsconfig.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (papb) [reply tsconfig.json:19]: I'm convinced!

File: test/storage/memory.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (papb) [reply test/storage/memory.test.ts:8]: It's still not fully clear to me - but instead of continuing the discussion here, I have opened an issue in your repo.

File: README.md [hunk 8]
```diff
@@ -315,7 +315,7 @@ class CustomStorage {
 const umzug = new Umzug({ storage: new CustomStorage(...) })
 ```
 
-If you're using TypeScript:
+Your instance must adhere to the [UmzugStorage](./src/storages/type-helpers/umzug-storage.ts) interface. If you're using TypeScript you can ensure this at compile time, and get IDE type hints by importing the `UmzugStorage` interface:
```
[the remark this answers is not in the document]
Comment (mmkal) [reply README.md:318]:
```suggestion
Your instance must adhere to the [UmzugStorage](./src/storage/contract.ts) interface. If you're using TypeScript you can ensure this at compile time, and get IDE type hints by importing it:
```

File: README.md [hunk 9]
```diff
@@ -315,7 +315,7 @@ class CustomStorage {
 const umzug = new Umzug({ storage: new CustomStorage(...) })
 ```
 
-Your instance must adhere to the [UmzugStorage](./src/storages/type-helpers/umzug-storage.ts) interface. If you're using TypeScript you can ensure this at compile time, and get IDE type hints by importing the `UmzugStorage` interface:
+Your instance must adhere to the [UmzugStorage](./src/storage/contract.ts) interface. If you're using TypeScript you can ensure this at compile time, and get IDE type hints by importing it:
```
Comment (papb) [inline README.md]: Nice catch on changing ```the `UmzugStorage` interface``` to `it`. Much better.

**Comment by papb** [comment]
Nice catch on changing ```the `UmzugStorage` interface``` to `it`. Much better.
